cpp/src/arrow/array/statistics.cc

```diff
@@ -15,7 +15,66 @@
 // specific language governing permissions and limitations
 // under the License.
 
-// This empty .cc file is for embedding not inlined symbols in
-// arrow::ArrayStatistics into libarrow.
-
 #include "arrow/array/statistics.h"
+
+#include <cmath>
+#include <type_traits>
+
+#include "arrow/compare.h"
+#include "arrow/util/logging_internal.h"
+namespace arrow {
+
+namespace {
+
+using ValueType = ArrayStatistics::ValueType;
+
+bool DoubleEquals(const double& left, const double& right, const EqualOptions& options) {
+  if (left == right) {
+    return options.signed_zeros_equal() || (std::signbit(left) == std::signbit(right));
+  } else if (options.nans_equal() && (std::isnan(left) || std::isnan(right))) {
+    return std::isnan(left) && std::isnan(right);
+  } else if (options.use_atol()) {
+    return std::fabs(left - right) <= options.atol();
+  } else {
+    return false;
+  }
+}
```

Comment:
Can we reuse existing implementation?

https://github.com/apache/arrow/blob/dfac0ccdba379692b29a74f09c62827df3ba2f3c/cpp/src/arrow/compare.cc#L72-L167